### PR TITLE
Android (Experimental support)

### PR DESCRIPTION
**Pull request by danielimprobable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
This contains the first experimental Android support for the GDK.

What is remaining:
* Support for AArch64
* Support for x86
* Streamlined workflow
* Docs

[Related Engine PR](https://github.com/improbableio/UnrealEngine/pull/57)

#### Tests
This has just been tested for local deployments and the workflow still needs more work. Expect further productisation PRs down the line.

#### Documentation
Docs are in the works and will be introduced in a later PR.

Work in progress doc https://docs.google.com/document/d/1s9dMXbdhFaZoBoinq_Ikm_rrMysqSgr9Qqt89fUtx1k/edit#

#### Primary reviewers
@m-samiec @joshuahuburn

### CODE REVIEW

File: Setup.bat [hunk 1]
```diff
@@ -67,19 +67,28 @@ call :MarkStartOfBlock "Create folders"
...
+    spatial package retrieve tools           schema_compiler-x86_64-win32                 %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\tools\schema_compiler-x86_64-win32.zip"
+    spatial package retrieve schema          standard_library                             %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\schema\standard_library.zip"
+    spatial package retrieve worker_sdk      c-dynamic-x86-msvc_md-win32                  %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\worker_sdk\c-dynamic-x86-msvc_md-win32.zip"
+    spatial package retrieve worker_sdk      c-dynamic-x86_64-msvc_md-win32               %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\worker_sdk\c-dynamic-x86_64-msvc_md-win32.zip"
+    spatial package retrieve worker_sdk      c-dynamic-x86_64-gcc_libstdcpp-linux         %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\worker_sdk\c-dynamic-x86_64-gcc_libstdcpp-linux.zip"
+    spatial package retrieve worker_sdk      c-dynamic-armeabi_v7a-clang_libcpp-android   %PINNED_CORE_SDK_VERSION%       "%CORE_SDK_DIR%\worker_sdk\c-dynamic-armeabi_v7a-clang_libcpp-android.zip"
```
Comment (joshuahuburn) [inline Setup.bat]: How much bloat do these add? Perhaps we want a --android flag in this setup script?

File: Setup.bat [hunk 2]
```diff
@@ -67,19 +67,28 @@ call :MarkStartOfBlock "Create folders"
...
+                        "Expand-Archive -Path \"%CORE_SDK_DIR%\worker_sdk\c-dynamic-x86_64-msvc_md-win32.zip\"              -DestinationPath \"%BINARIES_DIR%\Win64\" -Force; "^
+                        "Expand-Archive -Path \"%CORE_SDK_DIR%\worker_sdk\c-dynamic-x86_64-gcc_libstdcpp-linux.zip\"        -DestinationPath \"%BINARIES_DIR%\Linux\" -Force; "^
+                        "Expand-Archive -Path \"%CORE_SDK_DIR%\worker_sdk\c-dynamic-armeabi_v7a-clang_libcpp-android.zip\"  -DestinationPath \"%BINARIES_DIR%\Android\" -Force; "^
+                        "Expand-Archive -Path \"%CORE_SDK_DIR%\tools\schema_compiler-x86_64-win32.zip\"                     -DestinationPath \"%BINARIES_DIR%\Programs\" -Force; "^
+                        "Expand-Archive -Path \"%CORE_SDK_DIR%\schema\standard_library.zip\"                                -DestinationPath \"%BINARIES_DIR%\Programs\schema\" -Force;"
+
+                        REM "Expand-Archive -Path \"%CORE_SDK_DIR%\worker_sdk\c-dynamic-x86-android-clang_libcpp-android.zip\"  -DestinationPath \"%BINARIES_DIR%\Android\ARMx86\" -Force; "^
```
Comment (joshuahuburn) [inline Setup.bat]: Commends to be removed?

File: SpatialGDK/Extras/core-sdk.version [hunk 3]
```diff
@@ -1 +1 @@
-13.3.1
+13.4.0
```
Comment (joshuahuburn) [inline SpatialGDK/Extras/core-sdk.version]: Do our projects need an update with this?

File: SpatialGDK/Source/My_APL.xml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (joshuahuburn) [inline SpatialGDK/Source/My_APL.xml]: What does this file do and is it named correctly?

File: SpatialGDK/Source/Public/Schema/Component.h [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 #pragma once
 
+#include <utility>
```
Comment (joshuahuburn) [inline SpatialGDK/Source/Public/Schema/Component.h]: What's this needed for?

File: SpatialGDK/Source/SpatialGDK.Build.cs [hunk 6]
```diff
@@ -70,14 +70,30 @@ public SpatialGDK(ReadOnlyTargetRules Target) : base(Target)
                 ImportLibSuffix = ".lib";
                 SharedLibSuffix = ".dll";
                 break;
+            case UnrealTargetPlatform.Android:
+                LibPrefix = "lib";
+                ImportLibSuffix = SharedLibSuffix = ".so";
+                AdditionalPropertiesForReceipt.Add("AndroidPlugin", Path.Combine(ModuleDirectory, "My_APL.xml"));
+              //  string BuildPath = Utils.MakePathRelativeTo(ModuleDirectory, Target.RelativeEnginePath);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK.Build.cs]: Comments to be removed?

File: SpatialGDK/Source/SpatialGDK.Build.cs [hunk 7]
```diff
@@ -70,14 +70,30 @@ public SpatialGDK(ReadOnlyTargetRules Target) : base(Target)
...
         }
 
         string CoreSdkImportLib = System.String.Format("{0}worker{1}", LibPrefix, ImportLibSuffix);
         string CoreSdkSharedLib = System.String.Format("{0}worker{1}", LibPrefix, SharedLibSuffix);
 
-        PublicAdditionalLibraries.AddRange(new[] { Path.Combine(CoreSdkLibraryDir, CoreSdkImportLib) });
+        string libPath = Path.Combine(CoreSdkLibraryDir, CoreSdkImportLib);
+        if(Target.Platform == UnrealTargetPlatform.Android)
+        {
+            PublicAdditionalLibraries.AddRange(new[] { "worker" });
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK.Build.cs]: This is different for android?

File: Setup.bat (hunk 1 above)
Comment (danielimprobable) [reply Setup.bat]: They are about 5 mb each.  I think we are fine without the flag for now

File: SpatialGDK/Source/My_APL.xml (hunk 4 above)
Comment (danielimprobable) [reply SpatialGDK/Source/My_APL.xml]: Unreal uses this file to copy over binaries on the phone. Will rename the file

File: SpatialGDK/Source/SpatialGDK.Build.cs (hunk 7 above)
Comment (danielimprobable) [reply SpatialGDK/Source/SpatialGDK.Build.cs]: Yep

File: SpatialGDK/Extras/core-sdk.version (hunk 3 above)
Comment (danielimprobable) [reply SpatialGDK/Extras/core-sdk.version]: Yes they do, will create PRs for them

**Comment by Vatyx** [comment]
Shelving supporting Android for now.

**Pull request closed without merging**